feat: per-stage claude -p orchestration for implement-bead (#28)
scotthamilton77/agents-config: src/plugins/beads/.agents/skills/implement-bead/SKILL.md

name: implement-bead
description: >
  Use when a bead has the implementation-ready label and is ready for

>   autonomous execution. Drives ONE stage of the bead's molecule and exits.

Author: This frontmatter description seems to be going beyond the purpose of a skill description frontmatter.  Consult the docs for purpose and best practice, but understand the frontmatter's description purpose and write only to serve that purpose.

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

  Invoked by the shell driver via `claude -p /implement-bead <bead-id>`.
  The invoking agent is the ORCHESTRATOR only — all implementation work
  is done by subagents. Do NOT invoke this skill for beads that have not
  been through brainstorming (use start-bead to route correctly).
model: sonnet[1m]
effort: high
---

# implement-bead

> Drive ONE stage of the bead's molecule DAG and exit. The shell driver

Author: Too much explaining here.  Just explain what to do / how to perform the skill.  The coupling you have to bead-driver-test.sh is short-sighted noise.

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

(`scripts/bead-driver-test.sh`) calls this skill once per ready stage via
`claude -p --session-id <uuidv5> "/implement-bead <bead-id>"`. This skill
reads the current step, executes it, closes the step, and exits. The shell
driver handles looping — this skill MUST NOT loop to the next step.

## When to Use

> - Invoked by the shell driver via `claude -p /implement-bead <bead-id>`

Author: Too much information - why does the skill care whether it was invoked by claude -p or directly in an interactive claude session?

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

- The bead MUST have label `implementation-ready`

> - Do NOT invoke directly for beads that lack `implementation-ready` — use `start-bead`

Author: These are negative instructions to a human or an orchestrator, but buried in the "when to use" where these might be better expressed as constraints and then negatively described in a "gotchas" or "common mistakes" section.

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

done
```

### Step 4: Find the ONE Current Ready Step

Query the molecule for its current step using `--json` (text-mode is

> unreliable — see beads bug `2dx`):

Author: Skill should not tell the agent to go look at a bug issue.  Just say text mode is unreliable without the bug reference.

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

```bash
bd mol current <mol-id> --json
```

If no current step is returned (molecule is complete or all steps closed),

> EXIT immediately. The shell driver will handle close-walk and cleanup.

Author: ```suggestion
EXIT immediately. Do not close the bead or its ancestry.
```

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

> Do NOT squash or burn the molecule here — that is the driver's job.

Author: ```suggestion
Do NOT squash or burn the molecule here.
```

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

> Extract the step bead ID from the result.

Author: Isn't this only valid in the flow where there is an open step?

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

### Step 5: Execute the Step

Read the step bead's full description:

```bash

> bd show <step-bead-id>

Author: Does bd show also include labels, metadata, and other context associated with the bead?

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

```

Execute the instructions in that step bead. Dispatch subagents as needed
per the step description. The step description is self-contained — the
molecule's DAG enforces sequencing; you do not need to explain prior steps
to each new subagent.

**Subagent dispatch instructions** — for each step, pass:
1. The full text of the step bead description (from `bd show <step-bead-id>`)
2. The original bead context: title, AC, and bead ID
3. The cwd contract for this stage (decoded from the molecule's `worktree-path-*`
   label for most stages; repo root for `preflight` and `merge-or-handoff`)

> 4. The instruction: "Execute this step completely. Report back what you did."

Author: Consider having a report template

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

**Subagent isolation**: each subagent is independent and has no context
from previous subagents. The step bead description must be self-contained.

### Step 6: Close the Step and EXIT

> When the step is complete, close the step bead:

Author: If the subagent is sending back a report, the report could indicate that the step is not complete, e.g. a human escalation.  Let's be clear how the skill operator knows that the step is indeed complete.

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

```bash
bd close <step-bead-id> --reason "<brief summary of what was done>"
```

Then EXIT. Do NOT loop to the next step. Do NOT call `bd mol current` again.

> The shell driver polls for the next ready step and spawns a new `claude -p`

Author: Irrelevant to this skill

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

- Push branches
- Make git commits

All of these happen inside subagents executing the molecule step bead's instructions.

**One step per invocation.** This skill is called once per stage by the
shell driver. Driving multiple steps in a single invocation defeats the
per-stage context-bounding and session-id resumption model.

**No unauthorized merges.**
The implement-feature and fix-bug formulas end at `review-cycle`.
They do NOT merge. Merging requires explicit user authorization and is

> handled separately via the `merge-and-cleanup` formula.

Author: Fixed in d4d96db0d79d54641ab3d91d832e244d3b45a6d0. Updated SKILL.md No unauthorized merges section to name merge-or-handoff as terminal stage and clarify explicit authorization requirement.

- Push branches
- Make git commits

All of these happen inside subagents executing the molecule step bead's instructions.

> **One step per invocation.** This skill is called once per stage by the

Author: "...called once per stage by the hell driver" - this should be irrelevant, and as such it is noise.  Focus the skill on executing the step bead's scope only, and no other work.  Don't give the skill operator the need to worry about a driver or some kind of outer orchestration process.

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

shell driver. Driving multiple steps in a single invocation defeats the
per-stage context-bounding and session-id resumption model.

**No unauthorized merges.**
The implement-feature and fix-bug formulas end at `merge-or-handoff` (not
`review-cycle`). The `merge-or-handoff` step may pour the `merge-and-cleanup`
formula, but ONLY when the user has given explicit authorization ("go ahead and
merge", "ship it", etc.). Completing a PR or finishing the review cycle is NOT
authorization to merge. The `merge-and-cleanup` formula is the only authorized

> merge path — it requires explicit human sign-off before proceeding.

Author: Why does this skill need to know this?  If it does, you need to turn this into instructions on what to do in these cases.  Otherwise you're adding noise.

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

| Thought | Reality |
|---------|---------|
| "I'll loop to the next step before exiting" | No. One step per invocation. Exit after closing the step. The shell driver loops. |

> | "I'll invoke `ralf-it` / `superpowers:subagent-driven-development` directly" | No. Those methodology skills run INSIDE molecule steps, not as peers. |

Author: This is confusing, because step 5 executes a step, and since ralf-it and subagent-driven-development both require subagents, only the top-level agent following this skill could do these things.  Right?

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.

| "The step is small — I'll skip the subagent and do it in the main agent" | No. Main agent orchestrates, subagents implement. Dispatch even for small steps. |
| "I'll skip the formula and just run the work directly" | The formula IS the workflow. Skipping it skips the gate. |

> | "Brainstorming finished cleanly, so the user must want implementation" | No. Hand off to run-queue / shell driver by default. |

Author: This is irrelevant for this skill

Author: Acknowledged — the worker agent architecture (bead-implementor, implement-bead SKILL.md design, report contract) will be redesigned in a dedicated follow-up PR. This PR delivers the pipeline scaffolding; the worker layer redesign is next.
